asynciterator.ts

```diff
@@ -777,9 +784,121 @@ export class IntegerIterator extends AsyncIterator<number> {
   }
 }
 
+/**
+ * A synchronous mapping function from one element to another.
+ * The iterator performing the mapping is passed as a second argument.
+ */
+export type MapFunction<S, D = S, I extends AsyncIterator<D> = AsyncIterator<D>> =
+  (item: S, iterator: I) => D | null;
+
+/**
+  An iterator that calls a synchronous mapping function
+  on every item from its source iterator.
+  @extends module:asynciterator.AsyncIterator
+*/
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  protected _source: AsyncIterator<S>;
+  private readonly _destroySource: boolean;
+  private readonly _mappings: MapFunction<any, any, MappingIterator<S, D>>[];
+  private readonly _mappingRoot: InternalSource<any>;
+
+  // This is wrong: readable should be set by listening to source events
+  get readable() {
```

Comment:
This breaks things, because certain `readable` events will not be fired. We need to actually set `readable = true`.